org.spoofax.terms/src/org/spoofax/interpreter/terms/IStrategoArrayList.java

```diff
@@ -0,0 +1,234 @@
+package org.spoofax.interpreter.terms;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.RandomAccess;
+import org.spoofax.terms.StrategoArrayList;
+import org.spoofax.terms.StrategoArrayListBuilder;
+import org.spoofax.terms.StrategoArrayListIterator;
+import org.spoofax.terms.StrategoArrayListTail;
+import org.spoofax.terms.StrategoList;
+import org.spoofax.terms.TermList;
+import org.spoofax.terms.util.TermUtils;
+
+import static org.spoofax.terms.AbstractTermFactory.EMPTY_TERM_ARRAY;
+
+/**
+ * Most of the shared implementation for {@link StrategoArrayList} and {@link StrategoArrayListTail}
+ */
+public interface IStrategoArrayList extends IStrategoTerm, IStrategoList, RandomAccess, InternalIStrategoArrayList {
```

Comment:
This should be a class. Technically, it should extend `StrategoList`, but because code should not depend on the fact that a `IStrategoList` is a `StrategoList`, this is fine. But at least it should extend `StrategoTerm`. If this is a class, you don't need the `InternalIStrategoArrayList` interface.

Reply:
I don't want to extend `StrategoTerm` with `StrategoArrayListTail` but I do want to share code. That's what this interface does for me. I can't make it a class because StrategoArrayList still extends `StrategoTerm`, and it can't extend two classes.

Reply:
That would make `StrategoArrayListTail` extend `StrategoTerm`. I'm trying to avoid that, because it contains unnecessary fields and even some harmful functionality I don't want (such as `internalSetAnnotations`).

Reply:
Unused functionality does not matter for performance or memory usage, as you will just override it. And `StrategoTerm` has only these fields: the term's `annotations`, the cached `hashcode`, and the term's `attachment`. You have code that basically use the same fields as `hashcode` and `annotations`, leaving only the `attachment` field. Is saving the four or eight bytes of an unused field worth the complexity?

Reply:
I don't need annotations. I could cache the hashcode but currently don't. I don't need the attachments. But indeed, the big question of this PR is, is this complexity worth it? Sounds like you're opinion is "no". I'm starting to think the same.

Reply:
It feels like your original `StrategoArrayList` is already a huge improvement. This PR might be worth it, but without benchmarks we cannot know.